Post-processing

actyp opened the pull request:
The concept of this PR is to keep only the necessary functionality and reduce external dependencies from the normal usage of *ProtocolState-Fuzzer*. So I propose the following: 

* *ProtocolState-Fuzzer* outputs only the learned `DOT` model and does not attempt to convert it to `PDF` via the `dot` utility in system's PATH.

* The conversion, though, is handy to have available, so I created the `DotProcessor` class, which can be used from the user.

* The `LearnerResult` is enhanced with the corresponding `StateFuzzerEnabler` containing the configs used, which have useful information.

The `DotProcessor` class is supposed to be used after the learning on the data received from `LearnerResult`. This means that this class and other user-defined ones concern the post-processing of `LearnerResults`.

---
For example, this concept can be used to operate on the learned `DOT` model in order to convert it to `PDF` and consecutively beautify it (whatever this means in the user's context). 

```java
List<LearnerResult> results = commandLineParser.parse(args);
for (LearnerResult res : results) {
    CustomDotProcessor.exportToPDF(res);
    CustomDotProcessor.beautify(res);
}
```
For example, the `CustomDotProcessor.exportToPDF(LearnerResult)` function uses the `DotProcessor.exportToPDF(File)` function after checking if the provided `LearnerResult` is not null.

---
Essentially, *ProtocolState-Fuzzer* is concerned only with the output of the learned `DOT` model, without having any external dependency to `dot` utility and the post-processing of this learned `DOT` file is left to the user.

pfg666 commented:
I do see benefit in decoupling DOT post-processing from the core state fuzzing flow (separation of concerns). At the same time, I think there should be some default forms of post-processing built in (e.g., do nothing, or  generate PDF via GraphViz).  A state fuzzer may perform 'post-processing' of the generated models.  By default, this could be 'do nothing'. But a user may want to use a standard DOT -> PDF conversion. Or, they might want to implement their own, more advanced post-processing step. 

Rather than removing post-processing, I am wondering whether a better way would be to abstract it away, and make it configurable.

actyp commented:
I agree more with @pfg666, because my rationale is to decouple post-processing. 

What holds me back from abstracting is that an abstraction, i.e. an interface, limits the flexibility of the implementing class. 

For instance, if the interface provides the method
```
void process(LearnerResult result)
```
then it would not suffice for some use cases, where the 'processor' would need to return a value.

My main reason to leave post-processing entirely to the user (via the `LearnerResult`after parsing) is because it offers them the maximum flexibility and requires a minimum amount of additional code for the most common case (the call to `DotProcessor.exportToPDF(File)`), while providing total control of the post-processing.

This way the post-processing remains decoupled, optional and is controlled entirely by the user, while *ProtocolState-Fuzzer* provides (if needed) the necessary function for the most common case, which is the conversion to PDF.

actyp commented:
So shall we proceed with this or not ?

actyp commented:
I added the post-processing in `CommandLineParser` and now there are  three parse methods:

* `parse(String[])`, which parses and returns the results (the same as before)

* `parseAndExport(String[])`, which parses the arguments and then converts the generated DOT files to PDF. It returns the results of the parsing. (It is to be used in the common case)

* `parseAndConsume(String[], List<Consumer<LearnerResult>>)`, which parses the arguments and then consecutively applies each consumer to the results. It also returns the results of parsing.

Additional Changes:

* Added `DotProcessor.exportToPDF(LearnerResult)` function for convenience
* Enhanced the `LearnerResult` functions on `empty` and `normal` state

---

A usage example of the first parse function can be like this, in order to post-process the results manually:
```java
List<LearnerResult> results = commandLineParser.parse(args);

// post-process if needed
```

---

A usage example of the second parse function can be like this, in order to export the DOT file to PDF:
```java
List<LearnerResult> results = commandLineParser.parseAndExport(args);

// post-process if needed
```

---

A usage example for the third parse function can be like this, in order to first export the DOT file to PDF and then beautify the DOT file:
```java
List<Consumer<LearnerResult>> consumers = List.of(DotProcessor::exportToPDF, CustomDotProcessor::beautify);
List<LearnerResult> results = commandLineParser.parseAndConsume(args, consumers);

// additional post-process if needed
```

---

What do you think ?

kostis commented:
This option is better, IMO, but still I would prefer three `Parse` methods with one, two and three arguments.

Alternatively, if the above is confusing, we could have a single method `Parse(String[] args, bool exportPDF, List<Consumer<LearnerResult>> consumers)` where one is required to supply `false` and an empty list as the last two arguments if (s)he wants the functionality of the simplest `Parse`.

actyp commented:
OK, I implemented the three parse methods with one, two and three arguments.

kostis reviewed (approved):
Looks OK now, but it perhaps it should also come with some documentation.
(Update the `README` or add this somewhere else?)

actyp commented:
I updated the relevant section in the `README`.

kostis commented:
If @pfg666 does not have any other comments we can merge this.

pfg666 reviewed (approved):
Looks good to me!
